### ISSUE TITLE
Refactor Flags out of App.scala

### ISSUE THREAD

**Issue by frgomes** [twitter/util#210]
``App`` is opinionated in regards to command line parsing. Whilst this behavior is OK in the vast majority of use cases, there are circumstances where some more sophisticated command line parsing is desirable.

This issue is about allowing the developer choose whatever command line parser he/she wishes and finds more convenient for his/her requirements, instead of being locked into the way ``App`` decides to parse command line options.

### Expected behavior

There should be a way to obtain a "barebones version" of ``App`` so that this "new version" is absolutely oblivious about command line parsing. The developer should be able to retrieve ``args: Array[String]`` from the command line, unmodified, unprocessed, so that the developer decides and chooses whichever command line parser he/she finds more convenient for his/her specific needs.

### Actual behavior

At the moment, ``App.scala`` is opinionated about command line parsing.

**Comment by frgomes** [comment]
In my codebase I've "derived" ``AppLike`` from the original ``App`` like shown below.
I'm providing this code as a suggestion only. You guys are better skilled than me to decide whether such thing is acceptable and/or desirable.

```scala
package com.twitter.app

import com.twitter.conversions.time._
import com.twitter.util._
import java.lang.reflect.InvocationTargetException
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.atomic.AtomicReference
import java.util.logging.Logger
import scala.collection.JavaConverters._
import scala.collection.mutable

/**
 * A composable application trait that includes
 * basic application lifecycle (pre- and post- main).
 * Applications should be constructed
 * with modularity in mind, and common functionality should be
 * extracted into mixins.
 */
trait AppLike extends Closable with CloseAwaitably {

  /** The name of the application, based on the classname */
  val name: String = getClass.getName.stripSuffix("$")

  protected def exitOnError(reason: String): Unit = {
    exitOnError(reason, "")
  }

  /**
   * The details Fn may be an expensive operation (which could fail). We want to
   * ensure that we've at least written the `reason` field to System.err before
   * attempting to write the `detail` field so that users will at a minimum see the
   * `reason` for the exit regardless of any extra details.
   */
  protected def exitOnError(reason: String, details: => String): Unit = {
    System.err.println(reason)
    // want to ensure "reason" is written before attempting to write any details
    System.err.flush()
    System.err.println(details)
    Await.ready(close(), closeDeadline - Time.now)
    System.exit(1)
  }

  private val inits: mutable.Buffer[() => Unit] = mutable.Buffer.empty
  private val premains: mutable.Buffer[() => Unit] = mutable.Buffer.empty
  private val exits: ConcurrentLinkedQueue[Closable] = new ConcurrentLinkedQueue
  private val lastExits: ConcurrentLinkedQueue[Closable] = new ConcurrentLinkedQueue
  private val postmains: ConcurrentLinkedQueue[() => Unit] = new ConcurrentLinkedQueue

  // finagle isn't available here, so no DefaultTimer
  protected lazy val shutdownTimer: Timer = new JavaTimer(isDaemon = true)

  /**
   * Invoke `f` before anything else.
   */
  protected final def init(f: => Unit): Unit = {
    inits += (() => f)
  }

  /**
   * Invoke `f` right before the user's main is invoked.
   */
  protected final def premain(f: => Unit): Unit = {
    premains += (() => f)
  }

  /** Minimum duration to allow for exits to be processed. */
  final val MinGrace: Duration = 1.second

  /**
   * Default amount of time to wait for shutdown.
   * This value is not used as a default if `close()` is called without parameters. It simply
   * provides a default value to be passed as `close(grace)`.
   */
  def defaultCloseGracePeriod: Duration = Duration.Zero

  /**
   * The actual close grace period.
   */
  @volatile private[this] var closeDeadline = Time.Top

  /**
   * This is satisfied when all members of `exits` and `lastExits` have closed.
   *
   * @note Access needs to be mediated via the intrinsic lock.
   */
  private[this] var closing: Future[Unit] = Future.never

  /**
   * Close `closable` when shutdown is requested. Closables are closed in parallel.
   */
  final def closeOnExit(closable: Closable): Unit = synchronized {
    if (closing == Future.never) {
      // `close()` not yet called, safe to add it
      exits.add(closable)
    } else {
      // `close()` already called, we need to close this here, immediately.
      closable.close(closeDeadline)
    }
  }

  /**
   * Register a `closable` to be closed on application shutdown after those registered
   * via `closeOnExit`.
   *
   * @note Application shutdown occurs in two sequential phases to allow explicit
   *       encoding of resource lifecycle relationships. Concretely this is useful
   *       for encoding that a monitoring resource should outlive a monitored
   *       resource.
   *
   *       In all cases, the close deadline is enforced.
   */
  final def closeOnExitLast(closable: Closable): Unit = synchronized {
    if (closing == Future.never) {
      // `close()` not yet called, safe to add it
      lastExits.add(closable)
    } else {
      // `close()` already called, we need to close this here, but only
      // after `close()` completes and `closing` is satisfied
      closing
        .transform { _ =>
          closable.close(closeDeadline)
        }
        .by(shutdownTimer, closeDeadline)
    }
  }

  /**
   * Invoke `f` when shutdown is requested. Exit hooks run in parallel and are
   * executed after all postmains complete. The thread resumes when all exit
   * hooks complete or `closeDeadline` expires.
   */
  protected final def onExit(f: => Unit): Unit = {
    closeOnExit {
      Closable.make { deadline => // close() ensures that this deadline is sane
        FuturePool.unboundedPool(f).by(shutdownTimer, deadline)
      }
    }
  }

  /**
   * Invoke `f` after the user's main has exited.
   */
  protected final def postmain(f: => Unit): Unit = {
    postmains.add(() => f)
  }

  /**
   * Notify the application that it may stop running.
   * Returns a Future that is satisfied when the App has been torn down or errors at the deadline.
   */
  final def close(deadline: Time): Future[Unit] = synchronized {
    closing = closeAwaitably {
      closeDeadline = deadline.max(Time.now + MinGrace)
      val firstPhase = Closable
        .all(exits.asScala.toSeq: _*)
        .close(closeDeadline)
        .by(shutdownTimer, closeDeadline)

      firstPhase
        .transform { _ =>
          Closable.all(lastExits.asScala.toSeq: _*).close(closeDeadline)
        }
        .by(shutdownTimer, closeDeadline)
    }

    closing
  }

  def main(args: Array[String]): Unit = {
    try {
      nonExitingMain(args)
    } catch {
      case e: Throwable =>
        e.printStackTrace()
        exitOnError("Exception thrown in main on startup")
    }
  }

  def nonExitingMain(args: Array[String]): Unit = {
    registry.register(this)

    for (f <- inits) f()

    for (f <- premains) f()

    // Get a main() if it's defined. It's possible to define traits that only use pre/post mains.
    val mainMethod =
      try Some(getClass.getMethod("main"))
      catch { case _: NoSuchMethodException => None }

    // Invoke main() if it exists.
    mainMethod foreach { method =>
      try method.invoke(this)
      catch { case e: InvocationTargetException => throw e.getCause }
    }

    for (f <- postmains.asScala) f()

    // We discard this future but we `Await.result` on `this` which is a
    // `CloseAwaitable`, and this means the thread waits for the future to
    // complete.
    close(defaultCloseGracePeriod)

    // The deadline to 'close' is advisory; we enforce it here.
    Await.result(this, closeDeadline - Time.now)
  }

  val registry = AppLike
}

object AppLike {
  private[this] val log = Logger.getLogger(getClass.getName)
  private[this] val ref = new AtomicReference[Option[AppLike]](None)

    /**
   * The currently registered App, if any. While the expectation is that there
   * will be a single running App per process, the most-recently registered
   * App will be returned in the event that more than one exists.
   */
  def registered: Option[AppLike] = ref.get

  private[app] def register(app: AppLike): Unit =
    ref.getAndSet(Some(app)).foreach { existing =>
      log.warning(
        s"Multiple com.twitter.app.App main methods called. ${existing.name}, then ${app.name}"
      )
    }
}
```

This is an usage case in my codebase:
```scala
object Server {
  def apply(args: Args): Server = new Server(args)
}
class Server(val args: Args) extends ServerLike
trait ServerLike extends AppLike {

  val args: Args // command line already processed and exposed as a case class
...
}
```

**Comment by kevinoliver** [comment]
i get what you are going after.

i have an idea that would be a far smaller change, but may still get you what you want.

something like this. you'd add this to `App`:
```
  protected def parseFlags(): Unit = {
    // the current code...
    flag.parseArgs(args, allowUndefinedFlags) match {
      case Flags.Ok(remainder) =>
        _args = remainder.toArray
      case Flags.Help(usage) =>
        throw FlagUsageError(usage)
      case Flags.Error(reason) =>
        throw FlagParseException(reason)
    }  
}
```

and call it from `nonExitingMain`.

then your App implementation can do whatever it likes in terms of command line parsing (or none as it may be).

**Comment by smlance** [comment]
If I understand correctly, I think @kevinoliver's suggestion is pretty slick. Do you think this would suit your needs, @frgomes? If so, would you like to submit a PR? If you don't think you'll have time then one of us can merge this internally.

**Comment by frgomes** [comment]
@smlance : I'm sorry for the long silence. I'm going to work on this matter in the next few days. Thanks.

**Comment by cacoco** [comment]
@frgomes I've merged a change in cf7860e5c2037a496b596e1688bf5873875e3018 that hopefully addresses this issue. We'll be performing a release soon, please test it out and feel free to open a new ticket if there's still an issue. Thanks!
